File: src/Solutions/Euler080.hs
module Solutions.Euler080 where
import Utils.Numeric(binsearch)
import Data.List
import Data.Char

v = 10^101 :: Integer
findSqRoot :: Integer -> Integer
findSqRoot n = binsearch (\x -> x * x <= n * v * v) v (n * v)

numbers = [2..99] \\ map (^2) [2..9]
compute = sum $ map (sum . take 100 . map digitToInt . show . findSqRoot) numbers

euler080 :: IO String
euler080 = return $ show compute 

File: rust/src/solutions/euler080.rs
use crate::utils::numeric::binsearch::binsearch;
use crate::utils::numeric::digits::Digits;
use num::{cast::ToPrimitive, integer::Roots, pow::Pow, BigUint};

pub fn euler080() -> String {
    let v = &BigUint::from(10u32).pow(101u32);
    (2u32..100)
        .filter_map(|n| {
            if n.sqrt().pow(2) == n {
                None
            } else {
                let _n = &BigUint::from(n);
                binsearch(v, &(v * _n), |x: BigUint| &x * &x <= _n * v * v)
                    .digits()
                    .iter()
                    .take(100)
                    .sum::<BigUint>()
                    .to_usize()
            }
        })
        .sum::<usize>()
        .to_string()
}
